be/src/exec/pipeline/fragment_context.cpp

```diff
@@ -99,13 +101,40 @@ void FragmentContext::count_down_pipeline(RuntimeState* state, size_t val) {
 
     finish();
     auto status = final_status();
-    state->exec_env()->wg_driver_executor()->report_exec_state(query_ctx, this, status, true);
+    state->exec_env()->wg_driver_executor()->report_exec_state(query_ctx, this, status, true, true);
 
     destroy_pass_through_chunk_buffer();
 
     query_ctx->count_down_fragments();
 }
 
+bool FragmentContext::need_report_exec_state() {
+    auto* state = runtime_state();
+    auto* query_ctx = state->query_ctx();
+    if (!query_ctx->enable_profile()) {
+        return false;
+    }
+
+    return (MonotonicNanos() - _last_report_exec_state_ns) >= query_ctx->get_runtime_profile_report_interval_ns();
+}
+
+void FragmentContext::report_exec_state_if_necessary() {
+    auto* state = runtime_state();
+    auto* query_ctx = state->query_ctx();
+    if (!query_ctx->enable_profile()) {
+        return;
+    }
+    auto now = MonotonicNanos();
+    auto last_report_exec_state_ns = _last_report_exec_state_ns.load();
+    if (now - last_report_exec_state_ns < query_ctx->get_runtime_profile_report_interval_ns()) {
+        return;
+    }
+    if (_last_report_exec_state_ns.compare_exchange_strong(last_report_exec_state_ns, now)) {
+        LOG(ERROR) << "report runtime profile, query_id=" << print_id(_query_id);
+        state->exec_env()->wg_driver_executor()->report_exec_state(query_ctx, this, Status::OK(), false, true);
```

Comment:
Is there a possibility that a huge number of  report_exec_states rpc invocation in a very large cluster when users lanuch high concurrency queries overwhelms FE so that FE can not serve queries. so it would be better that runtime profile only support in such scenarios:
1. users designate enable_profile in SET_VAR explicilty.
2. ETL queries, such as CTAS, ITAS and etc.

Reply:
I think the first one is already supported, check the above condition
```cpp
    if (!query_ctx->enable_profile()) {
        return;
    }
```

And the second, there will always be long-running querys at all situations. So we cannot restrict this features only to some of the scenerios.

Reply:
If the query finishes within the scheduled time, which is defined by the `runtime_profile_report_interval`, this feature will not perform additional state reporting.

Reply:
What I mean is forbidding runtime profile in other scenarios except above two ones, is it reasonable?

Reply:
I don't think so. For a large-scale data scenario, queries may potentially last for several tens of minutes.

And another point is: there's no overhead for queries with shorter processing times, because the additional state report only happens if the query lasts for more than `runtime_profile_report_interval` seconds.